internal/handlers/common/filter.go

```diff
@@ -348,32 +352,38 @@ func filterFieldExpr(doc *types.Document, filterKey string, expr *types.Document
 
 		case "$gt":
 			// {field: {$gt: exprValue}}
-			switch exprValue.(type) {
-			case *types.Document, *types.Array:
-			case types.Regex:
+			if _, ok := exprValue.(types.Regex); ok {
 				msg := fmt.Sprintf(`Can't have RegEx as arg to predicate over field '%s'.`, filterKey)
 				return false, NewCommandErrorMsgWithArgument(ErrBadValue, msg, exprKey)
-			default:
-				if arrValue, ok := fieldValue.(*types.Array); ok {
-					// Filter the array by only keeping the same type as exprValue.
-					// This is because array comparison with greater than compares
-					// against the maximum value of the same type from the array.
-					// All numbers are treated as the same type.
-					// Example:
-					// expr {v: {$gt: 42}}
-					// value [{v: 40}, {v: 41.5}, {v: "foo"}, {v: nil}]
-					// Above compares the maximum number of array 41.5 to the filter 42,
-					// and results in Less. Other values "foo" and nil which are
-					// not number type are not considered for $gt comparison.
-					arr := arrValue.FilterArrayByType(exprValue)
-
-					if arr.Len() == 0 {
-						// The array does not contain any element with the same type as exprValue.
-						return false, nil
-					}
+			}
 
-					fieldValue = arr.Max()
+			if ok := filterByComparisonOperandApplicable(fieldValue, exprValue); !ok {
```

Comment:
See https://www.mongodb.com/docs/manual/reference/method/db.collection.find/#type-bracketing
> The document with _id equal to "avocados" is not returned because its qty value is of type string while the [$gt](https://www.mongodb.com/docs/manual/reference/operator/query/gt/#mongodb-query-op.-gt) operand is of type integer.